sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,396 @@
+performed during a full compilation from source to final executable.  The
+compiler driver is responsible for creating the fat object and the
+`clang-linker-wrapper` tool is responsible for the general functionality that is
+performed during the link.
+
+![High level view of the offloading flow](images/OffloadGeneralFlow.svg
+ "General Offload Flow")
+
+*Diagram 1: General Offload Flow*
+
+More specific details of the clang-linker-wrapper behavior can be seen in the
+[clang-linker-wrapper](#clang-linker-wrapper) section.
+
+## Fat Binary Representation
+
+Binaries generated during the offload compilation will be 'bundled' together
+to create a conglomerate fat binary.  Depending on the type of binary, the
+packaging of the device section will be different.
+
+## Device only Binaries
+
+Generation and consumption of device only binaries is also available.  The
+device only binaries are represented by the file generated from the
+[Packager](#packager) to generate an 'Offload Binary'.
+
+## Fat Binary Generation
+
+The generation of the fat binary will take place in the driver.  The model from
+the community is generating the fat binary as a secondary process when creating
+the host object.  Generation will be separated from the host compilation step.
+This is being done to enable proper support for using an external host compiler
+as well as taking advantage of potential parallelism during compilation of both
+the host and target device binaries.
+
+The fat object in the new model is generated during the host compilation.
+The host compilation takes an additional argument which points to the device
+binary which will be embedded in the final object.  Generation will be
+separated out to allow for potential parallelism during compilation of both the
+host and target device binaries.
+
+![Creating the fat object](images/OffloadFatObject.svg
+ "Fat Object Generation")
+
+*Diagram 2: Fat Object Generation*
+
+### Packager
+
+When dealing with multiple device binaries, an additional step is performed to
+package the multiple device binaries before being added to the host object.
+This additional step is performed with the `clang-offload-packager` taking
+image inputs containing information relating to the target triple,
+architecture setting and offloading kind.
+
+The `clang-offload-packager` is run during 'fat object' generation regardless
+of the number of device binaries being added to the conglomerate fat object.
+The device binaries are contained in what is designated as an ‘Offload Binary’.
+These binaries can reside in a variety of binary formats including Bitcode
+files, ELF objects, executables and shared objects, COFF objects, archives or
+simply stored as an offload binary.
+
+We should have the ability to package SPIR-V based device binaries in the
+offload section of any given binary.  These device binaries will be packaged
+as normal with the packager and placed within the given section.
```

Comment:
Can we add an example usage of clang-offload-packager here?

Thanks

Reply:
simple example added (command line usage)